scenario1: implement testnet validation step 1.2

tmigone opened the pull request:
This PR implements step 1.2 on the L2 validation plan with a script plus a test file that can be run to assert the scenario.

>This scenario sets up and tests the following:
>
>- two indexers with stake
>- each indexer has two or three allocations, one of which is open
>- four subgraphs, with 0, 400, 10000 and 12500 tokens signalled, with signal coming from three different curators

abarmat commented on e2e/scenarios/open-allocations.ts:
```diff
@@ -0,0 +1,49 @@
+
+  // == Fund participants
+  console.log('\n== Fund indexers')
+  await fundAccountsEth(deployer, indexers, fund.ethAmount)
```
Maybe I'd put `fund` amounts inside the Indexer fixture, like a `.grtBalance` and `.ethBalance` within each indexer, and maybe rename `fund` to `default` as placeholder?

tmigone replied:
good idea

abarmat commented on scripts/e2e:
```diff
@@ -25,7 +25,9 @@ npx hardhat migrate:unpause --network localhost --graph-config config/graph.loca
 # Run tests using the localhost evm instance and the localhost graph config
 npx hardhat e2e --network localhost --graph-config config/graph.localhost.yml
-
+npx hardhat e2e:scenario create-subgraphs --network localhost --graph-config config/graph.localhost.yml
```
```
  if (chainId == 1337) {
    await (cli.wallet.provider as providers.JsonRpcProvider).send('evm_setAutomine', [false])
  }
```
That line after `migrate` finishes is making the rest of the tests to run slow. What if we set automine:true by default for local in the hardhat.config?

This is the current config:
```
      mining: {
        auto: false,
        interval: 13000,
      },
```

And then we set to false for specific tests. Seems to me that most of the time we want it to be fast during tests.

tmigone replied:
I think I agree with this. With the intent of being less disruptive for now I've added a parameter to the `migrate` call so we have the flexibility to leave auto mine on or off after the deployment is done (and assuming we are in 1337 ofc).

abarmat commented on e2e/scenarios/lib/subgraph.ts:
```diff
@@ -1,17 +1,35 @@
+    await contracts.GNS.nextAccountSeqID(publisher.address),
   )
-  await tx.wait()
+  await sendTransaction(publisher, contracts.GNS, 'publishNewSubgraph', [
```
I'd like to eventually move to using the more expressive: 

`const tx = await contracts.GNS.connect(publisher).publishNewSubgraph(deploymentId, randomHexBytes(), randomHexBytes())` 

And then have a wait tx helper that prints debugging. We could work on that on a different PR (in the future ™️).

tmigone replied:
yes this would be awesome, tracking here: https://github.com/graphprotocol/contracts/issues/702

abarmat reviewed (commented):
It is looking good, I left a few comments.

tmigone commented:
@abarmat I'll work on addressing your comments but I think you didn't review the latest commit, I think you were on `6087247` (see https://github.com/graphprotocol/contracts/pull/670/commits). I only noticed because there are two later commits that refactor the scenario1 into more generic and re-usable parts.

tmigone commented:
@abarmat, I'll reply to your comments now. Note that you didn't review the latest commit but an intermediate one. Some of your comments were addressed by later commits. 

In any case, since this PR is L1 exclusive and not using any of the new L1 features (reservoir, gateway) I'm cherry picking these commits into `dev` on a separate PR so I'll close this one.

tmigone commented on TESTING.md:
```diff
@@ -70,6 +71,25 @@ npx hardhat e2e:scenario <scenario> --network <network> --graph-config config/gr
+
+#### Scenario 1
+
+This scenario sets up and tests the following:
```
Good idea!

tmigone commented on tasks/e2e/e2e.ts:
```diff
@@ -29,19 +27,6 @@ task('e2e', 'Run all e2e tests')
-    }
   })
 
 task('e2e:config', 'Run deployment configuration e2e tests')
```
The TESTING.MD file describe the differences but in a nutshell:

**Config**
Configuration are tests that evaluate parameters that don't change with protocol usage. Some examples:
- default reserve ratio for curation matching whatever is on the config file
- Rewards manager being a minter

These can be run against any protocol deployment (mainnet, goerli, localhost, etc) and they should pass if the network is healthy. 

**Init**
Init tests on the other hand test parameters that are meant to change over time (as the protocol is being used). For example:
- totalSupply of the token being 10B. This changes with new issuance.
- staking contract allowance for allocation exchange being max uint256. This changes as vouchers are redeemed. 

These tests only make sense to run on a fresh deployment where no actions have been taken.

tmigone commented on e2e/scenarios/lib/subgraph.ts:
```diff
@@ -1,17 +1,35 @@
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
+import { BigNumber } from 'ethers'
+import { solidityKeccak256 } from 'ethers/lib/utils'
 import { NetworkContracts } from '../../../cli/contracts'
```
Yes agreed, I already started putting some things in `common-ts` but it's a slow process

tmigone commented on e2e/scenarios/lib/staking.ts:
```diff
@@ -1,18 +1,59 @@
+    'allocateFrom',
+    [indexer.address, subgraphDeploymentID, amount, allocationId, metadata, proof],
+    {
+      gasLimit: 2000000,
```
I ran into problems when running on Goerli, the automatic gas estimation was not making a great job and the transactions were reverting with out of gas error. I then saw we have this gasLimit value set for some operations in the CLI.

tmigone commented on e2e/scenarios/lib/staking.ts:
```diff
@@ -1,18 +1,59 @@
+
+  console.log(`\nClosing ${allocationId}...`)
+  await sendTransaction(indexer, contracts.Staking, 'closeAllocation', [allocationId, poi], {
+    gasLimit: 2000000,
```
See other comment.

tmigone commented on e2e/scenarios/lib/curation.ts:
```diff
@@ -1,15 +1,21 @@
-  await tx.wait()
+  console.log(`\nAdd ${amount} in signal to subgraphId ${subgraphId}..`)
+  await sendTransaction(curator, contracts.GNS, 'mintSignal', [subgraphId, amount, 0], {
+    gasLimit: 2000000,
```
Same, see other comment.

tmigone commented:
Commits from this PR have been cherry picked into https://github.com/graphprotocol/contracts/pull/703
Closing PR.
